File: cpp/src/checkarg.cpp

#include "checkarg.hpp"
#include "checkarg_private.hpp"
#include <sstream>
#include <iostream>
using namespace std;

/**
 * \class CheckArg
 * \brief a lightweight and easy to use command line parser
 *
 * It's inspired by ArgumentParser from python,
 * therefore, if you've used that one before, you'll find this one quite similar.
 * It is, intentionally, in no way a complete reimplementation or port of the former.
 *
 * The main goal is for it to be easy to understand and use.
 * Secondly, its intended to be non-intrusive and lightweight.
 * Therefore it pulls no 'external' dependencies and relies completely on the STL.
 *
 * That way it's pretty easy to integrate into an existing project,
 * because it's non-intrusive, you dont need to 'build your project arount it'.
 *
 * Commandline options use the common so-called GNU style,
 * e.g. use double dashes and look something like these: <br>
 *  -\-option[=value] <br>
 *  -\-option [value] <br>
 * Additionally, abreviations of those in so-called POSIX style can be added.
 * They look like this: <br>
 *  -o [value]
 * I'll call them long and short options, respectively.
 *
 * Using CheckArg as a developer can be divided in two phases, pre-parse() and post-parse().
 * In the pre-parse phase, you'll define your commandline options using the add() members.
 * In the post-parse phase, you can check, which options where given
 * and get their values, if they've got one.
 *
 * Callbacks will, for now, be called while parsing,
 * if an option has no value, immediately after it was parsed,
 * otherwise, immediately after the value was parsed.
 *
 * Maybe they'll be split in immediate and deferred ones later,
 * but defferred ones would currently have no advantage over doing<br>
 * <code>  if( ca.isset("option") ) doOption();</code><br> after parse(),
 * so in favor of being lightweight, they're not implementet for now.
 */


map <int,string>
CheckArgPrivate::errors = {
	{CA_ALLOK,  "Everything is fine"},
	{CA_ERROR,  "An Error occurred"},
	{CA_INVARG, "Unknown command line argument"},
	{CA_INVVAL, "Value given to non-value argument"},
	{CA_MISSVAL, "Missing value of argument"},
	{CA_CALLBACK, "Callback returned with error code"},
};

// c'tors

/**
 * \brief constructs a CheckArg object
 * \param argc the number of arguments, you'll want to pass argc gotten from main()
 * \param argv the list of arguments, you'll want to pass argv from main()
 * \param appname the name of your application, will be used in the usage-line
 */
CheckArg::CheckArg(const int argc, char ** argv, const std::string &appname)
  : p(new CheckArgPrivate(this, argc,argv,appname)){
}
/**
 * \brief constructs a CheckArg object
 * \param argc the number of arguments, you'll want to pass argc gotten from main()
 * \param argv the list of arguments, you'll want to pass argv from main()
 * \param appname the name of your application, will be used in the usage-line
 * \param desc description of your programm, used in help
 */
CheckArg::CheckArg(const int argc, char ** argv, const std::string &appname, const std::string &desc)
  : p(new CheckArgPrivate(this, argc, argv, appname, desc)){
}
/**
 * \brief constructs a CheckArg object
 * \param argc the number of arguments, you'll want to pass argc gotten from main()
 * \param argv the list of arguments, you'll want to pass argv from main()
 * \param appname the name of your application, will be used in the usage-line
 * \param desc description of your programm, used in help
 * \param appendix text which will be added at the end of help
 */
CheckArg::CheckArg(const int argc, char ** argv, const std::string &appname, const std::string &desc, const std::string &appendix)
  : p(new CheckArgPrivate(this, argc,argv,appname,desc,appendix)){
}

// private c'tors
CheckArgPrivate::CheckArgPrivate(CheckArg* ca, const int argc, char** argv, const std::string &appname)
	: argc(argc), argv(argv), parent(ca), appname(appname), autohelp_on(false),
	  usage_line(appname + " [options]") {}

CheckArgPrivate::CheckArgPrivate(CheckArg* ca, const int argc, char** argv, const std::string &appname, const std::string &desc)
	: argc(argc), argv(argv), parent(ca), appname(appname), descr(desc), autohelp_on(false),
	  usage_line(appname + " [options]") {}

CheckArgPrivate::CheckArgPrivate(CheckArg* ca, const int argc, char** argv, const std::string &appname, const std::string &desc, const std::string &appendix)
	: argc(argc), argv(argv), parent(ca), appname(appname), descr(desc), appendix(appendix), autohelp_on(false),
	  usage_line(appname + " [options]") {}


/**
 * \brief set help's text for positional arguments
 * \param usage text to be appended to the usage line, like " [files...]"
 * \param descr text describing the positional arguments, will be used in help
 */
void
CheckArg::set_posarg_help(const std::string &usage, const std::string &descr ) {
  p->posarg_help_usage = usage;
  p->posarg_help_descr = descr;
}


/**
 * \brief set a completely own usage line
 * \param str the usage line to use
 */
void
CheckArg::set_usage_line(const std::string &str) {
  p->usage_line = str;
}


/**
 * \brief get argv[0], the programms "name"
 * \return argv[0]
 */
std::string
CheckArg::argv0() {
  return p->argv[0];
}

/**
 * \brief get all positional arguments
 *
 * all arguments not being options like '-h' or '-\-help'
 * or values of the former are considered positional arguments
 *
 * For Example:<br>
 *  <code>  ./ca_user pos0 -h -i input.txt pos1 -t pos2 pos3</code><br>
 *  all arguments starting with pos are positional arguments,
 *  given, that '-i' has a value and '-t' and '-h' have not
 * \return vector of positional arguments
 */
std::vector<std::string>
CheckArg::pos_args() const {
  return p->pos_args;
}


/**
 * \brief get error message for given errno
 * \param errno error number
 * \return error message string
 * \see CAError
 */
std::string
CheckArg::str_err(const int errno){ return CheckArgPrivate::errors[errno]; }

void
CheckArgPrivate::ca_error(int eno, const std::string &info, ...) const {
#ifdef CA_PRINTERR
  va_list al;
  va_start(al, info);
  char *buff;
  int i=vasprintf(&buff,info.c_str(),al);
  if (i > -1){
    std::cerr << "Error: " << errors[eno] << buff << std::endl;
    free(buff);
  }
#endif
}

// bigger functions
//


/**
 * \brief add a command line option the parser shall accept
 * \param sopt the short option, like 'h' for '-h'
 * \param lopt the long option, like 'help' for '-\-help'
 * \param help a short description for the option, used in generated help
 * \param has_val does the option have a value
 * \return a return code from CAError
 * \see CAError
 */
int CheckArg::add(const char sopt, const std::string &lopt, const std::string &help, const bool has_val){
	p->valid_args[lopt] = { has_val, sopt, help };
	p->short2long[sopt] = lopt;
	return CA_ALLOK;
}

/**
 * \brief add a command line option the parser shall accept
 * \param sopt the short option, like 'h' for '-h'
 * \param lopt the long option, like 'help' for '-\-help'
 * \param cb a callback to be called whenever the option is encountered
 * \param help a short description for the option, used in generated help
 * \param has_val does the option have a value
 * \return a return code from CAError
 * \see CAError
 */
int CheckArg::add(const char sopt, const std::string &lopt,
                  std::function<int(CheckArgPtr,const std::string &, const std::string &)> cb,
                  const std::string &help, const bool has_val){
	p->valid_args[lopt].has_val = has_val;
	p->valid_args[lopt].sopt    = sopt;
	p->valid_args[lopt].help    = help;
	p->valid_args[lopt].cb      = cb;
	p->short2long[sopt]         = lopt;
	return CA_ALLOK;
}

/**
 * \brief add a command line option the parser shall accept
 * \param lopt the long option, like 'help' for '-\-help'
 * \param help a short description for the option, used in generated help
 * \param has_val does the option have a value
 * \return a return code from CAError
 * \see CAError
 */
int CheckArg::add(const std::string &lopt, const std::string &help, const bool has_val){
	p->valid_args[lopt].has_val = has_val;
	p->valid_args[lopt].help    = help;
	return CA_ALLOK;
}

/**
 * \brief add a command line option the parser shall accept
 * \param lopt the long option, like 'help' for '-\-help'
 * \param cb a callback to be called whenever the option is encountered
 * \param help a short description for the option, used in generated help
 * \param has_val does the option have a value
 * \return a return code from CAError
 * \see CAError
 */
int CheckArg::add(const std::string &lopt,
                  std::function<int(CheckArgPtr,const std::string &, const std::string &)> cb,
                  const std::string &help, const bool has_val){
	p->valid_args[lopt].has_val = has_val;
	p->valid_args[lopt].cb      = cb;
	p->valid_args[lopt].help    = help;
	return CA_ALLOK;
}


/**
 * \brief add auto generated '-\-help' message
 * \return CA_ALLOK
 */
int CheckArg::add_autohelp(){
	p->valid_args["help"].has_val = false; // add --help with no value
	p->valid_args["help"].sopt    = 'h';
	p->valid_args["help"].help    = "show this help message and exit";
	p->valid_args["help"].cb      = checkarg::show_autohelp; // set the autohelp callback
	p->short2long['h'] = "help";   // add -h mapped to --help
	p->autohelp_on = true;         // switch autohelp to on
	return CA_ALLOK;
}

/**
 * \brief parse the commnd line, this should come after any add() or add_autohelp()
 * \return CA_ALLOK on success, some other code from CAError otherwise
 * \see CAError
 */
int CheckArg::parse(){
	int ret = CA_ALLOK;
	for(int i=1; i<p->argc; ++i){ // start with 1 here, because argv[0] is special
		ret = p->arg(p->argv[i]);
		if( ret != CA_ALLOK ) break;
	}
	if( ! p->next_is_val_of.empty() ){
		p->ca_error(CA_MISSVAL, ": "+ string(p->argv[p->argc-1]) + "!");
		return CA_MISSVAL;
	}

	// free strings not necessary anymore, e.g. those for '--help'
	p->appname.clear();
	p->appendix.clear();
	p->usage_line.clear();
	p->descr.clear();
	p->posarg_help_descr.clear();
	p->posarg_help_usage.clear();
	p->next_is_val_of.clear();
	for( auto arg : p->valid_args)
		arg.second.help.clear();

	return ret;
}


/**
 * \brief get the value of a given option
 * \warning you shouldn't call this before parse()!
 * \param arg the long name of the option to get the value of
 * \return the value
 */
string CheckArg::value(const std::string &arg) const{

	auto pos = p->valid_args.find(arg);
	if( pos != p->valid_args.end() ){
		return pos->second.value;
	}
	return "";
}

/**
 * \brief check if an option was given on the command line
 * \param arg the option specified by it's long name
 * \return true if give, false otherwise
 */
bool CheckArg::isset(const std::string &arg) const {
	auto pos = p->valid_args.find(arg);
	return pos != p->valid_args.end() && !(pos->second.value.empty());
}

int CheckArgPrivate::arg(const std::string &arg){
	if( ! next_is_val_of.empty() ){
		// _next val of should be an opt with value
		//static_assert( _valid_args[_next_is_val_of].has_val );

		valid_args[next_is_val_of].value = arg;
		auto ret = call_cb(next_is_val_of);
		next_is_val_of.clear();
		return ret;
	}

	if( arg[0] == '-' ){ // it's an arg
		if( arg[1] == '-' ) { // it's a long one
			return arg_long(arg.substr(2));
		}

		// it's a short one or a group of short ones
		return arg_short(arg.substr(1));
	}

	// it's some positional arg
	pos_args.push_back(arg);
	return CA_ALLOK;
}

int CheckArgPrivate::arg_long(const std::string &arg){
	auto eqpos = arg.find('=');
	string real_arg, val;
	if( eqpos != string::npos ){
		real_arg = arg.substr(0, eqpos);
		val = arg.substr(eqpos+1);
	} else {
		real_arg = arg;
	}

	auto pos = valid_args.find(real_arg);
	if( pos != valid_args.end() ){
		if( pos->second.has_val && !val.empty() ) {
			// arg has value defined, and value is given by '='
			valid_args[real_arg].value = val;
		} else if( pos->second.has_val ){
			// value of arg is the next arg, remember that for the next call of arg
			next_is_val_of = real_arg;
		} else { // there's no value defined by add()
			if( !val.empty() ){
				// error?
				ca_error(CA_INVVAL, ": --" + real_arg + "!");
				return CA_INVVAL;
			}
			valid_args[real_arg].value = "x"; // mark arg as seen
		}

		if( !pos->second.has_val || !val.empty()) {
			// if arg has no val, or val is found already, call callback now, if there's one
			return call_cb(real_arg);
		}
		return CA_ALLOK;
	} else {
		ca_error(CA_INVARG, ": --" + real_arg + "!");
		return CA_INVARG;
	}
}

int CheckArgPrivate::arg_short(const std::string &arg){
	size_t len = arg.size();
	for(int i=0; i < len; ++i){
		auto pos = short2long.find(arg[i]);
		if( pos != short2long.end() ){ // there is such a short arg registered
			if( valid_args[pos->second].has_val ) { // if has val,
				if( i < len-1 ){ //remainder is interpreted as val,
					valid_args[pos->second].value = arg.substr(i+1);
					return call_cb(pos->second);
				} else { // or next_arg is treated as val
					next_is_val_of = pos->second;
				}
				return CA_ALLOK; // no further looping, we're done.
			} else {
				valid_args[pos->second].value = "x"; // mark arg as seen
				auto ret = call_cb(pos->second);
				if(ret != CA_ALLOK )
				  return ret;
			}
		} else {
			ca_error(CA_INVARG, string(": -") + arg[i] + "!");
			return CA_INVARG;
		}
	}
	return CA_ALLOK;
}

int CheckArgPrivate::call_cb(const std::string &arg){
	auto cbpos = valid_args.find(arg);
	if( cbpos != valid_args.end() && cbpos->second.cb ){
		int cbret = cbpos->second.cb(CheckArgPtr(parent), arg, cbpos->second.value);
		if(cbret != CA_ALLOK) {
			// if callback returns anything other than CA_ALLOK, there's been an error
			ca_error(CA_CALLBACK, ": %d!", cbret);
			return CA_CALLBACK;
		}
	}
	return CA_ALLOK;
}


int checkarg::show_autohelp(CheckArgPtr ca, const std::string&, const std::string &val){
	stringstream ss;
	size_t space = 0;
	for( auto &kv : ca->p->valid_args )
		space = max(space, kv.first.size() );

	space += 2; // add 2 more spaces

	ss << "Usage: " << ca->p->usage_line << " " << ca->p->posarg_help_usage << endl;

	if(!ca->p->descr.empty()) ss << endl << ca->p->descr << endl;

	ss << endl << "Options:" << endl;
	for(auto it=ca->p->valid_args.begin(); it != ca->p->valid_args.end(); ++it){
		auto sarg = ca->p->valid_args.find(it->first);
		if(sarg != ca->p->valid_args.end()) ss << "   -" << sarg->second.sopt << ",";
		else ss << "      ";
		ss << " --" << it->first << string(space-it->first.size(), ' ')
		   << ca->p->valid_args[it->first].help << endl;
	}
	if(!ca->p->posarg_help_descr.empty())
		ss << endl
			 << "Positional Arguments:" << endl
			 << ca->p->posarg_help_descr << endl;
	if(!ca->p->appendix.empty()) ss << endl << ca->p->appendix << endl;

	cout << ss.str() << flush;

	exit(0); // always exit after showing help
}

File: java/CheckArg.java
import java.util.*;
import java.util.concurrent.*;

public class CheckArg {

	public CheckArg(String args[], final String appname) {
		_args = args;
		_appname = appname;
		_autohelp_on = false;
		_usage_line = _appname + " [options]";

		_valid_args      = new HashMap<String,Boolean>();
		_valid_args_cb   = new HashMap<String,Callable<Void>>();
		_valid_args_vals = new HashMap<String,String>();
		_autohelp        = new HashMap<String,String>();
		_short2long      = new HashMap<Character,String>();
		_next_is_val_of  = "";
		_pos_args        = new ArrayList<String>();

		_descr = "";
		_appendix = "";
		_posarg_help_descr = "";
		_posarg_help_usage = "";
	}
	public CheckArg(String args[], final String appname, final String desc ) {
		this(args,appname);
		_descr = desc;
	}
	public CheckArg(String args[], final String appname, final String desc, final String appendix ) {
		this(args,appname,desc);
		_appendix = appendix;
	}

	public int add(final char sopt, final String lopt, final String help){
		return add(sopt,lopt,help,false);
	}
	public int add(final char sopt, final String lopt, final String help, final boolean has_val){
		_valid_args.put(lopt, has_val);
		_autohelp.put(lopt, help);
		_short2long.put(sopt,lopt);
		return Error.CA_ALLOK.ordinal();
	}

	public int add(final char sopt, final String lopt, final String help, final Callable<Void> cb){
		return add(sopt,lopt,help,cb,false);
	}
	public int add(final char sopt, final String lopt, final String help, final Callable<Void> cb, final boolean has_val){
		add(sopt, lopt, help, has_val);
		_valid_args_cb.put(lopt,cb);
		return Error.CA_ALLOK.ordinal();
	}

	public int add(final String lopt, final String help){
		return add(lopt,help,false);
	}
	public int add(final String lopt, final String help, final boolean has_val){
		_valid_args.put(lopt, has_val);
		_autohelp.put(lopt, help);
		return Error.CA_ALLOK.ordinal();
	}

	public int add(final String lopt, final String help, final Callable<Void> cb){
		return add(lopt,help,cb,false);
	}
	public int add(final String lopt, final String help, final Callable<Void> cb, final boolean has_val){
		add(lopt,help,has_val);
		_valid_args_cb.put(lopt, cb);
		return Error.CA_ALLOK.ordinal();
	}

	public int add_autohelp(){
		add('h', "help", "show this help message and exit", new ShowAutohelp(), false);
		_autohelp_on = true;
		return 0;
	}

	public int parse(){
		int ret;
		
		for (String a : _args){
			ret = arg(a);
			if( ret != Error.CA_ALLOK.ordinal() )
				return ret;
		}
		if( ! _next_is_val_of.isEmpty() ){
			ca_error(Error.CA_MISSVAL, ": "+_args[_args.length-1]+"!");
			return Error.CA_MISSVAL.ordinal();
		}
		return Error.CA_ALLOK.ordinal();
	}

	public void set_posarg_help(final String usage, final String descr){
		_posarg_help_descr = descr;
		_posarg_help_usage = usage;
	}

	public void set_usage_line(final String usage){
		_usage_line = usage;
	}

	public String argv0(){
		return _appname;
	}

	public List<String> pos_args(){
		return _pos_args;
	}

	public String value(final String arg){
		return _valid_args_vals.get(arg);
	}

	public boolean isset(final String arg){
		return _valid_args_vals.get(arg) != null;
	}

	public static String str_err(final Error errno){
		return _errors.get(errno);
	}
	public static void ca_error(Error eno, final String info){
		System.err.println("Error: "+_errors.get(eno.ordinal())+(info==null?"!":info));
	}

	private class ShowAutohelp implements Callable<Void> {
		public Void call(){
			int space = 0;
			for( String l : _valid_args.keySet() )
				space = Math.max(space, l.length());
			space += 2; // 2 more spaces for gap

			System.out.println("Usage: " + _usage_line + " " + _posarg_help_usage);

			if( ! _descr.isEmpty() )
				System.out.println("\n"+_descr);

			System.out.println("\nOptions:");
			for ( String l : _valid_args.keySet() ) {
				String sarg = long2short(l);
				if(!sarg.isEmpty())
					System.out.print("   -"+sarg+",");
				else
					System.out.print("      ");

				System.out.println(" --"+l+str_repeat(' ', space-l.length())+_autohelp.get(l));
			}

			if( ! _posarg_help_descr.isEmpty() ){
				System.out.println("\nPositional Arguments:");
				System.out.println(_posarg_help_descr);
			}

			if( ! _appendix.isEmpty() ){
				System.out.println("\n"+_appendix);
			}

			System.exit(0);

			return null; // always exit after showing help
		}

	}
	
	private String str_repeat(char c, int times){
		char[] chars = new char[times];
		Arrays.fill(chars, c);
		return new String(chars);
	}

	private int arg(final String arg){
		if(! _next_is_val_of.isEmpty() ){
			_valid_args_vals.put(_next_is_val_of, arg);
			int ret = _call_cb(_next_is_val_of);
			_next_is_val_of = "";
			return ret;
		}

		if( arg.charAt(0) == '-' ) { // yay, some arg
			if(arg.charAt(1) == '-' ) // it's a long one
				return arg_long(arg.substring(2));

			// short one or group of short ones
			return arg_short(arg.substring(1));
		}

		// some positional arg
		_pos_args.add(arg);
		return Error.CA_ALLOK.ordinal();
	}

	private int arg_long(final String arg){
		int eqpos = arg.indexOf('=');
		String real_arg, val;
		if( eqpos == -1 ){
			real_arg = arg;
			val = "";
		} else {
			real_arg = arg.substring(0,eqpos);
			val = arg.substring(eqpos+1);
		}

		Boolean hasval = _valid_args.get(real_arg);
		if( hasval != null ){
			if( hasval.booleanValue() && !val.isEmpty() ){
				// has val and val is given after an '='
				_valid_args_vals.put(real_arg, val);
			} else if( hasval.booleanValue()) {
				// value of arg is the next arg, remember that for the next call of arg
				_next_is_val_of = real_arg;
			} else { // arg has no value 
				if( ! val.isEmpty() ){
					ca_error(Error.CA_INVVAL, ": --"+real_arg+"!");
					return Error.CA_INVVAL.ordinal();
				}
				_valid_args_vals.put(real_arg, "x"); // arg was seen
			}

			if( ! hasval.booleanValue() || ! val.isEmpty()){
				// if arg has no val, or val is found already, call callback now, if there's one
        return _call_cb(real_arg);
			}

			return Error.CA_ALLOK.ordinal();
		}
		ca_error(Error.CA_INVARG, ": --"+real_arg+"!");
		return Error.CA_INVARG.ordinal();
	}

	private int arg_short(final String arg){
		int len = arg.length();

		for (int i = 0; i < len; i++) {
			String larg = _short2long.get(arg.charAt(i));
			if( larg != null ){
				if( _valid_args.get(larg).booleanValue() ){ // if has val
					if( i < len-1 ){
						_valid_args_vals.put(larg, arg.substring(i+1));
					} else { // next arg is val
						_next_is_val_of = larg;
					}
					return Error.CA_ALLOK.ordinal();
				} else {
					_valid_args_vals.put(larg, "x");
					int ret = _call_cb(larg);
					if( ret != Error.CA_ALLOK.ordinal() )
					  return ret;
				}
			} else {
				ca_error(Error.CA_INVARG, ": -"+arg.charAt(i)+"!");
				return Error.CA_INVARG.ordinal();
			}
		}
		return Error.CA_ALLOK.ordinal();
	}
	
  private int _call_cb(final String larg){
		Callable<Void> cb = _valid_args_cb.get(larg);
		if( cb != null ){
			try{
				cb.call();
			}catch(Exception e){
				ca_error(Error.CA_CALLBACK, ": "+e+"!");
				e.printStackTrace();
				return Error.CA_CALLBACK.ordinal();
			}
		}
		return Error.CA_ALLOK.ordinal();
  }

	private String long2short(final String larg){
		for (Map.Entry<Character,String> e : _short2long.entrySet() ) {
			if( e.getValue().equals(larg) )
				return Character.toString(e.getKey());
		}
		return null;
	}

	private String _appname, _descr, _appendix, _usage_line, _posarg_help_descr, _posarg_help_usage;
	private boolean _autohelp_on;
	private String _next_is_val_of;

	private Map<String,Boolean>        _valid_args;
	private Map<String,String>         _valid_args_vals;
	private Map<String,Callable<Void>> _valid_args_cb;
	private Map<String,String>         _autohelp;
	private Map<Character,String>      _short2long;
	private List<String>               _pos_args;
	private String                     _args[];

	private static final Map<Error, String> _errors;
	static {
		_errors = new HashMap<Error,String>();
		_errors.put(Error.CA_ALLOK,    "Everything is fine");
		_errors.put(Error.CA_ERROR,    "An Error occurred");
		_errors.put(Error.CA_INVARG,   "Unknown command line argument");
		_errors.put(Error.CA_INVVAL,   "Value given to non-value argument");
		_errors.put(Error.CA_MISSVAL,  "Missing value of argument");
		_errors.put(Error.CA_CALLBACK, "Callback returned with error code");
	}

	public enum Error {
		CA_ALLOK,
		CA_ERROR,
		CA_INVARG,
		CA_INVVAL,
		CA_MISSVAL,
		CA_CALLBACK,
	}
}
